Panic (`byte index 60 is out of bounds`) when trying to save new file on remote host

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

In a remote project, I open a new buffer, immediately try and save it, and as soon as I type a single character, the application crashes.

I attached logs for my latest crash below, but after reproducing it a few times, the path that throws the panic is sometimes different (3 out of 6 are the one I've attached). Another example is: `byte index 6 is out of bounds of `tf``. In all cases, the index is always 4 above the length of the string.

I created an empty project with just 2 directories and 1 file in each and it didn't crash. The crash is happening in a large (private company) monorepo.

### Environment

Zed: v0.150.3 (Zed Preview)
OS: macOS 14.6.1
Memory: 16 GiB
Architecture: aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary><pre>
<!-- Click below this line and paste or drag-and-drop your log-->
2024-08-26T21:04:20.222139Z [ERROR] {
  "thread": "main",
  "payload": "byte index 60 is out of bounds of `config/src/services/external-secrets/helm-builds/prod/au`",
  "location_data": {
    "file": "crates/ui/src/components/label/highlighted_label.rs",
    "line": 75
  },
  "backtrace": [
    "<backtrace::capture::Backtrace>::create",
    "<backtrace::capture::Backtrace>::new",
    "zed::reliability::init_panic_hook::{closure#0}",
    "std::panicking::rust_panic_with_hook",
    "std::panicking::begin_panic_handler::{{closure}}",
    "std::sys_common::backtrace::__rust_end_short_backtrace",
    "_rust_begin_unwind",
    "core::panicking::panic_fmt",
    "core::str::slice_error_fail_rt",
    "core::str::slice_error_fail",
    "ui::components::label::highlighted_label::highlight_ranges",
    "<file_finder::new_path_prompt::NewPathDelegate as picker::PickerDelegate>::render_match",
    "gpui::elements::uniform_list::uniform_list::<&str, gpui::elements::div::Stateful<gpui::elements::div::Div>, picker::Picker<file_finder::new_path_prompt::NewPathDelegate>, <picker::Picker<file_finder::new_path_prompt::NewPathDelegate>>::render_element_container::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::uniform_list::UniformList as gpui::element::Element>::prepaint::{closure#1}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::uniform_list::UniformList as gpui::element::Element>::prepaint::{closure#1}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::uniform_list::UniformList as gpui::element::Element>::prepaint::{closure#1}>::{closure#0}",
    "<gpui::elements::uniform_list::UniformList as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::uniform_list::UniformList> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::element::Drawable<gpui::element::AnyElement> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::view::AnyView> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::element::Drawable<gpui::view::View<workspace::status_bar::StatusBar>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::view::AnyView> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::window::WindowContext>::draw",
    "<gpui::window::Window>::new::{closure#4}::{closure#1}",
    "<gpui::window::Window>::new::{closure#4}",
    "gpui::platform::mac::window::step",
    "<gpui::platform::mac::platform::MacPlatform as gpui::platform::Platform>::run",
    "<gpui::app::App>::run::<zed::main::{closure#3}>",
    "zed::main",
    "std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>",
    "std::rt::lang_start::<()>::{closure#0}",
    "std::rt::lang_start_internal",
    "_main"
  ],
  "app_version": "0.150.3",
  "release_channel": "Zed Preview",
  "os_name": "macOS",
  "os_version": "14.6.1",
  "architecture": "aarch64",
  "panicked_on": 1724706260222,
  "installation_id": "1004f7d9-7049-4f23-80cb-c122cd01e50e",
  "session_id": "b1df2cfa-37f8-4ffd-8e2d-5cda55cd92ed"
}
<!-- Click above this line and paste or drag-and-drop your log--></pre></details>

--- a/crates/file_finder/src/new_path_prompt.rs
+++ b/crates/file_finder/src/new_path_prompt.rs
@@ -107,8 +107,10 @@ impl Match {
 
         if let Some(path_match) = &self.path_match {
             text.push_str(&path_match.path.to_string_lossy());
+            let mut whole_path = PathBuf::from(path_match.path_prefix.to_string());
+            whole_path = whole_path.join(path_match.path.clone());
             for (range, style) in highlight_ranges(
-                &path_match.path.to_string_lossy(),
+                &whole_path.to_string_lossy(),
                 &path_match.positions,
                 gpui::HighlightStyle::color(Color::Accent.color(cx)),
             ) {

